### PR TITLE
update terraform code to version 0.12

### PR DESCRIPTION
**Pull request by venturel**

### CODE REVIEW

**Review by Komorebi-E — APPROVED** [review]
As SM, LGTM. E.g. follows the guidelines for Splat, e.g. `aws_eip.nat_gateway[count.index].id`

File: nat_gateway/main.tf [hunk 1]
```diff
@@ -1,24 +1,33 @@
+terraform {
```
Comment (iuriaranda) [inline nat_gateway/main.tf]: Can you use a `versions.tf`  file? 😛

File: nat_gateway/output.tf [hunk 2]
```diff
@@ -1,7 +1,8 @@
 output "ids" {
-  value = ["${aws_nat_gateway.gateway.*.id}"]
+  value = [aws_nat_gateway.gateway.*.id]
```
Comment (iuriaranda) [inline nat_gateway/output.tf]: you don't need the list notation here, as `aws_nat_gateway.gateway.*.id` is already a list

File: subnets/outputs.tf [hunk 3]
```diff
@@ -1,3 +1,4 @@
 output "ids" {
-  value = ["${aws_subnet.subnets.*.id}"]
+  value = [aws_subnet.subnets.*.id]
```
Comment (iuriaranda) [inline subnets/outputs.tf]: Same here about the list notation

File: vpc/main.tf [hunk 4]
```diff
@@ -1,94 +1,107 @@
...
-# Create a VPC
 resource "aws_vpc" "main" {
-  cidr_block           = "${var.cidr_block}"
+  cidr_block           = var.cidr_block
   enable_dns_support   = true
   enable_dns_hostnames = true
 
-  tags = "${merge("${merge(var.extra_tags_vpc,var.tags)}",map("Name", "${var.project} ${var.environment} VPC", "Environment", "${var.environment}", "Project", "${var.project}"))}"
+  tags = merge(
+    merge(var.extra_tags_vpc, var.tags),
```
Comment (iuriaranda) [inline vpc/main.tf]: you can simplify this in a single merge function

```
tags = merge(
  var.extra_tags_vpc,
  var.tags,
  {
    "Name"        = "${var.project} ${var.environment} VPC"
    "Environment" = var.environment
    "Project"     = var.project
  },
)
```

File: vpc/outputs.tf [hunk 5]
```diff
@@ -1,44 +1,45 @@
...
 output "private_management_subnets" {
   description = "List of the private_management subnets id created"
-  value       = "${module.private_management_subnets.ids}"
+  value       = module.private_management_subnets.ids
 }
 
 output "private_rts" {
   description = "List of the ids of the private route tables created"
-  value       = ["${aws_route_table.private.*.id}"]
+  value       = [aws_route_table.private.*.id]
```
Comment (iuriaranda) [inline vpc/outputs.tf]: Same here and the next output about the list notation

**Review by iuriaranda — CHANGES_REQUESTED** [review]
Some details

**Review by iuriaranda — APPROVED** [review]

**Review by iuriaranda — CHANGES_REQUESTED** [review]
The readme should be updated as well with the variable type changes and the examples

File: subnets/variables.tf [hunk 6]
```diff
@@ -41,10 +42,11 @@ variable "num_subnets" {
 }
 
 variable "route_tables" {
-  type    = "list"
+  type    = list(string)
   default = []
 }
 
 variable "num_route_tables" {
   default = "0"
```
Comment (iuriaranda) [inline subnets/variables.tf]: The type of this variable should be enforced to a `number` as well

File: nat_gateway/variables.tf [hunk 7]
```diff
@@ -1,17 +1,17 @@
 variable "private_route_tables" {
-  type = "list"
+  type = list(string)
```
Comment (iuriaranda) [inline nat_gateway/variables.tf:2]: The type changes should also be reflected in the readme documentation

**Review by iuriaranda — APPROVED** [review]

**Pull request merged**